File: Battleship/src/main/java/de/htwg/battleship/actor/childactor/ShipActor.java
// ShipController.java
package de.htwg.battleship.actor.childactor;

import akka.actor.ActorRef;
import akka.actor.Props;
import akka.actor.UntypedActor;
import akka.pattern.Patterns;
import akka.util.Timeout;
import de.htwg.battleship.actor.childactor.grandchild.BorderTrueActor;
import de.htwg.battleship.actor.childactor.grandchild.CollisionActor;
import de.htwg.battleship.actor.messages.BorderMessage;
import de.htwg.battleship.actor.messages.CollisionMessage;
import de.htwg.battleship.actor.messages.ShipMessage;
import de.htwg.battleship.model.IPlayer;
import de.htwg.battleship.model.IShip;
import scala.concurrent.Await;
import scala.concurrent.Future;

import java.util.LinkedList;
import java.util.List;
import java.util.concurrent.TimeUnit;

/**
 * ShipController to place the ships and test if there are collisions. to test
 * that it uses another controller which uses a chain-of-responsibility-pattern
 *
 * @author Moritz Sauter (SauterMoritz@gmx.de)
 * @version 1.00
 * @since 2014-11-27
 */
public class ShipActor extends UntypedActor {

    private static final Timeout TIMEOUT = new Timeout(3, TimeUnit.SECONDS);

    /**
     * Controller with a chain of responsibility.
     */
    private final ActorRef collisionActor =
        getContext().actorOf(Props.create(CollisionActor.class), "Collider");

    /**
     * Controller with a chain of responsibility.
     */
    private final ActorRef borderActor =
        getContext().actorOf(Props.create(BorderTrueActor.class), "Borderer");

    /**
     * Method to place a Ship on a field.
     *
     * @param ship      ship which should be placed
     * @param player    on which board the ship should be placed
     * @param boardSize
     *
     * @return true if the ship was placed false if there was a collision or the
     * ship is out of the field
     */
    public final boolean placeShip(final IShip ship, final IPlayer player,
                                   int boardSize) {
        Future<Object> futureBorder = Patterns
            .ask(borderActor, new BorderMessage(ship, boardSize), TIMEOUT);
        List<Future<Object>> futureCollision = playerShip(ship, player);
        boolean result = false;
        try {
            result = (boolean) Await.result(futureBorder, TIMEOUT.duration());
            for (Future<Object> collision : futureCollision) {
                if (!result) {
                    break;
                }
                result = !(boolean) Await.result(collision, TIMEOUT.duration());
            }
            if (result) {
                player.getOwnBoard().addShip(ship);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }

        return result;
    }

    /**
     * Utility-Method to set a ship for a specified player.
     *
     * @param ship   to place
     * @param player which player
     *
     * @return true if it is set, false if not
     */
    private List<Future<Object>> playerShip(final IShip ship,
                                            final IPlayer player) {
        List<Future<Object>> futureList = new LinkedList<>();
        IShip[] shipList = player.getOwnBoard().getShipList();
        for (int i = 0; i < player.getOwnBoard().getShips(); i++) {
            Future<Object> future = Patterns
                .ask(collisionActor, new CollisionMessage(ship, shipList[i]),
                     TIMEOUT);
            futureList.add(future);
        }
        return futureList;
    }

    @Override
    public void onReceive(Object message) throws Exception {
        if (!(message instanceof ShipMessage)) {
            unhandled(message);
            return;
        }
        ShipMessage shipMessage = (ShipMessage) message;
        boolean result =
            placeShip(shipMessage.getShip(), shipMessage.getPlayer(),
                      shipMessage.getBoardSize());
        getSender().tell(result, getSelf());
    }
}

File: Battleship/src/main/scala/de/htwg/battleship/actor/ShipActor.scala
package de.htwg.battleship.actor

import akka.actor.{Actor, ActorRef, Props}
import akka.pattern.ask
import akka.util.Timeout
import de.htwg.battleship.actor.messages.{BorderMessage, CollisionMessage, ShipMessage}
import de.htwg.battleship.model.{IPlayer, IShip}

import scala.collection.mutable.ListBuffer
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future
import scala.concurrent.duration._

object ShipActor {
    val ACTOR_NAME = "shipper"
}

class ShipActor extends Actor {
    val borderer = context.actorOf(Props[BorderActor], BorderActor.ACTOR_NAME)
    val collider = context.actorOf(Props[CollisionActor], CollisionActor.ACTOR_NAME)
    implicit val timeout = Timeout(5 seconds)

    def createColliderFutures(ship: IShip, player: IPlayer): List[Future[Boolean]] = {
        var list = new ListBuffer[Future[Boolean]]()
        val shipList = player.getOwnBoard.getShipList
        for (i <- 0 until player.getOwnBoard.getShips) {
            list += (collider ? CollisionMessage(ship, shipList(i))).mapTo[Boolean]
        }
        list.toList
    }

    // TODO: think about an abstract
    def reduce(one: Boolean, two: Boolean): Boolean = one && two

    def placeShip(ship: IShip, player: IPlayer, boardSize: Int, ref: ActorRef) = {
        val borderFuture = (borderer ? BorderMessage(ship, boardSize)).mapTo[Boolean]
        val collisionFutures = createColliderFutures(ship, player)
        val allFutures = List(collisionFutures, List(borderFuture)).flatten
        val aggrFuture = Future sequence allFutures
        aggrFuture onSuccess {
            case results: List[Boolean] =>
                ref ! results.reduce(reduce)
        }
    }

    /**
      * Method to check if a ship could validly placed on the board of a player
      *
      * @return
      */
    override def receive: Receive = {
        case msg: ShipMessage =>
            placeShip(msg.ship, msg.player, msg.boardSize, sender())
        case msg => unhandled(msg)
    }
}
